### PR TITLE
[New `classes` concept exercise]: ellens-alien-game

### PR DESCRIPTION
**Pull request by PaulT89**
**_BethanyG's Note:_**  This PR is for a concept exercise introducing `classes`.  See https://github.com/exercism/python/issues/2349 for more details/discussion.

Introduces the concept of classes. 
~~Introduction.md is empty~~
~~Design.md is empty~~

(_Update: Other Notes_)
I should probably also mention that there are no 'General' hints yet.  Not sure if that's important.

(_edited to add checklist_)

- [x] config.json & UUID for exercise
- [x]  introduction.md
- [x]  design.md
- [x]  instructions.md
- [x]  hints.md
- [x]   links.json
- [x]  .meta/exemplar.py
- [x]  Added exercise icon to .meta/config.json file
- [x]  Formatted .meta/config.json file
- [x]  classes_test.py (_tests_)
- [x]  classes.py (_stub file_)
- [x]  Formatted `config.json` file
- [ ]  Spelling, language, and punctuation (_**needs a final check**_)

### CODE REVIEW

File: exercises/concept/ellens-alien-game/alien_class.py
Comment (IsaacG) [inline exercises/concept/ellens-alien-game/alien_class.py:10]: I'm not sure which docstring style is being followed here. The numpy style has the summary/description before the param/return list.

File: exercises/concept/ellens-alien-game/alien_class.py
Comment (IsaacG) [inline exercises/concept/ellens-alien-game/alien_class.py:11]:
```suggestion
    instance per position.
```

File: exercises/concept/ellens-alien-game/alien_class_test.py
Comment (IsaacG) [inline exercises/concept/ellens-alien-game/alien_class_test.py:9]: This comment should be a docstring and should terminate in a period.

File: exercises/concept/ellens-alien-game/alien_class_test.py
Comment (IsaacG) [inline exercises/concept/ellens-alien-game/alien_class_test.py:29]: Use docstrings to document methods, here and below.

**Comment by BethanyG** [comment]
@IsaacG   -- This PR was a WIP, and should not have been reviewed for things like punctuation, etc.  In future, please wait for a maintainer to go through a PR for content or subject matter before nit-picking language, formatting or style.  Reviewing for detail before content adds noise and confusion as to what should be done to make a larger PR (_like a new concept exercise_) ready for release.  Thanks!

File: exercises/concept/ellens-alien-game/.docs/instructions.md
Comment (PaulT89) [reply exercises/concept/ellens-alien-game/.docs/instructions.md:44]: Thanks!

File: exercises/concept/ellens-alien-game/.meta/exemplar.py
Comment (PaulT89) [reply exercises/concept/ellens-alien-game/.meta/exemplar.py:11]: Ellen's the one who will be using it, and she doesn't care. 😃  But for real though, with the inclusion of the `is_alive` method, I guess the decision ultimately comes down to where you want the `if` statement to be.  The alternative methods would be the following:

```python
################snip##############

Either way, I suppose I should rework the alien_class_test file.
    def hit(self):
        self.health -= 1 if self.health > 0 else 0

    def is_alive(self):
        return bool(self.health)

################snip##############
```

File: exercises/concept/ellens-alien-game/.docs/instructions.md
Comment (PaulT89) [reply exercises/concept/ellens-alien-game/.docs/instructions.md:92]: True.  However, the point here is to be as clear as possible as to what input should generate what output, and I worry that a list may add an unnecessary layer of obfuscation (at least to the eyes of a beginner).

Specifically:
- The following example of expected input/output would make it unclear as to what the `total_aliens_created` variable is supposed to be (i.e. a class variable) and how it is supposed to work (i.e. increment upon instantiation):

```python
>>> aliens = [Alien(5, 1), Alien(3, 0)]
>>> aliens[0].total_aliens_created
2
>>> aliens[1].total_aliens_created
2
```

- The following example of expected input/output would make it more clear as to what the `total_aliens_created` variable is supposed to be/do, but is functionally indistinguishable from the original:

```python
>>> aliens = [Alien(5, 1)]
>>> aliens[0].total_aliens_created
1
>>> aliens.append(Alien(3, 0))
>>> aliens[1].total_aliens_created
2
>>> aliens[0].total_aliens_created
2
```

The above expected IO example may also raise a question in the user's mind as to why someone would go to the effort of doing such a thing and not simply get the length of the list instead.  However, if there are two obviously separate variables, then it might occur to the user that multiple alien objects can be bound to individual variables, or split among multiple data structures, yet with each still knowing how many have been created without having to consolidate them all.  

It essentially shows the user a neat trick that only classes can do.  (Well, strictly speaking, I suppose you _could_ do it without classes, but only with a lot of pain and effort.)

**Review by BethanyG — COMMENTED** [review]
@PaulT89  -- I think this is an **excellent** first pass!  🌟   I am going stop with the edits/feedback for this round - The GitHub UI is getting flakey on me, and I think there is a lot already here for you to process.  The test file still needs some additional feedback -- but it can wait until we get some other things sorted.  Thank you again for all of your hard work on this -- it's great.  Let me know if you have questions or issues - or if you need anything.

File: exercises/concept/ellens-alien-game/.meta/exemplar.py
Comment (PaulT89) [reply exercises/concept/ellens-alien-game/.meta/exemplar.py:11]: Okay.  I've fixed the test file so the user can do it whatever way they like and the test file won't fail them.

File: exercises/concept/ellens-alien-game/.docs/instructions.md
Comment (PaulT89) [reply exercises/concept/ellens-alien-game/.docs/instructions.md:27]: Keeping the first bit as it is (those are back-quotes around hit, not... er... quote-quotes - it's `hit`, not 'hit', because it's supposed to be in the code.  Then again, maybe I'm going overboard with the back-quotes.)

Changing the second bit to your suggestion.

File: exercises/concept/ellens-alien-game/alien_class_test.py
Comment (PaulT89) [reply exercises/concept/ellens-alien-game/alien_class_test.py:5]: Yeah, I noticed that the concept test files had that decorator, while the practice files didn't.  They also seemed to be laid out kind of differently, which is more or less why I mentioned my test file's layout.  I think I was using a practice test layout rather than a concept test layout.

File: exercises/concept/ellens-alien-game/.docs/instructions.md
Comment (PaulT89) [reply exercises/concept/ellens-alien-game/.docs/instructions.md:11]: Yeah, I don't know either.  When I was writing it up, I didn't want to be _so_ explicit about _what_ to do that I might accidentally stumble into telling the user _how_ to do it.  I do hint at the x and y coords being instance variables in the hints file, though maybe not explicitly enough.

The whole `x_coordinate` and `y_coordinate` thing is a good idea.  I didn't know that there was a single-letter-variable rule (whoops!), but I do see that explicitly specifying them as coordinates is probably going to help.  It will mean some finding/replacing, though 😆 

I like the new line 11 though.

File: exercises/concept/ellens-alien-game/.docs/instructions.md
Comment (PaulT89) [reply exercises/concept/ellens-alien-game/.docs/instructions.md:45]: I like it, except for the back-quotes around both 'is_alive' and 'method'.  Same reason as above.

File: exercises/concept/ellens-alien-game/.docs/instructions.md
Comment (PaulT89) [reply exercises/concept/ellens-alien-game/.docs/instructions.md:62]: Doh! 🤦  I called it a function instead of a method.  Also, same hesitation about the back-quotes.  Also (also), are the new x and y going to be `coordinate_x` and `coordinate_y` because before it was `x_coordinate` and `y_coordinate`?

File: exercises/concept/ellens-alien-game/.docs/instructions.md
Comment (PaulT89) [reply exercises/concept/ellens-alien-game/.docs/instructions.md:76]: Again, hesitant about the back-quotes around method and class, but otherwise yes.

File: exercises/concept/ellens-alien-game/.docs/instructions.md
Comment (PaulT89) [reply exercises/concept/ellens-alien-game/.docs/instructions.md:102]: Hmm... good point.

File: exercises/concept/ellens-alien-game/.docs/instructions.md
Comment (PaulT89) [reply exercises/concept/ellens-alien-game/.docs/instructions.md:115]: Yes, except for the conversation about which way around the `x`/`y` and `coordinate` goes

File: exercises/concept/ellens-alien-game/.docs/introduction.md
Comment (PaulT89) [reply exercises/concept/ellens-alien-game/.docs/introduction.md:3]: Sounds good. 👍

File: exercises/concept/ellens-alien-game/.meta/exemplar.py
Comment (PaulT89) [reply exercises/concept/ellens-alien-game/.meta/exemplar.py:18]: Again, good apart from whether the letter and 'coordinate' need to be switched here or at the beginning.

**Comment by IsaacG** [comment]
So ... can community members comment on this PR? Is there a specific signal we should wait for? Or is community reviews discouraged?

**Comment by BethanyG** [comment]
@IsaacG  -- Community _**reviews**_ of PRs are generally discouraged.  That doesn't mean no comments or discussion on them -- but that **does** mean that formally reviewing (_or appearing to formally review_) something and implying edits, approval, or disapproval for merging into the track is discouraged.  Especially if that review doesn't cover the entire scope of the PR requirements, or creates confusion on the part of the PR submitter.

Track content, style, and curriculum is the responsibility of the maintainers, who also have write access to the track and approval/disapproval for merging.  While discussion is encouraged in issues and on Slack, PRs are somewhat more formal.  
It's important that there be a (small) set of people who are aligned with regards to standards and decision making, and who make "final calls" on what is going to be accepted or rejected.

**Comment by iHiD** [comment]
Without trying to create a meta discussion here (and sorry @PaulT89 for doing so!), I just wanted to respond to the last couple of comments (hey @IsaacG and @BethanyG :))

So there's a specific challenge for PRs that create Concept Exercises, where its absolutely essential to get the exercise correctly aligned to the place and scope of the track, and all of the many many (sorry!) rules around Concept Exercises correctly worked through. In reality, that means that exercises get almost entirely rewritten 2 or 3 times before they land at a place where they're good enough for what I'd class as a "normal" review. There's also often files that overlap (an exercise + concept's intros, abouts, etc) where getting one of those perfect first is really essential before branching out too far. So TL;DR; a first-pass review by a maintainer often takes 2 or 3 hours and results in dozens or sometimes hundreds of comments. So what I really encourage is having one senior maintainer really owning that process, committing the time to working with the PR creator, and them as a pair working to get the exercise to a good place. At that stage, everyone's exhausted and merging that first pass down with `status: 'wip'` is the right thing to do.

Getting feedback from our top mentors is then absolutely vital, before the exercise goes live. So what I'd like to then happen is for the exercise to go out as a WIP exercise, which high-rep users see and get to test, and where they can then add extra issues or PRs about. That last bit is still flakey atm (as WIP exercises don't appear unless you go looking for them), but it's the right way to go about it in the long run, as it removes the burden on the original contributor/maintainer team getting everything perfect, but doesn't lead to huge PR cycles with multiple voices (which are really hard to manage).

On other PRs, it varies more. In general, I feel like all PRs should have one primary reviewer (whichever maintainer takes it on), and other maintainers and/or community members should act in a secondary role. Generally, I think that means that that secondary role has two really helpful points:
1. Proofreading spelling/grammar etc **once the first pass from the primary maintainer has happened**. The reason I put that last bit in bold is that it's really frustrating for contributors if they get a spelling/grammar review, fix it, and then a maintainer comes along and asks them to make more fundamental changes.
2. Pointing out things **in a non-actionable way, or to the maintainer**. It's brilliant when someone highlights things that I as the primary reviewer should think about or have missed, but do so in such a way that it doesn't immediately ask for changes from the author. So writing comments aimed at stating an opinion, or writing questions can work really well. For example, I was a secondary reviewer on a Ruby PR this week. My initial instinct was that the PR shouldn't be merged, but I didn't want to start that conversation with the author of the PR because that might give them mixed messages and be confusing for them, so instead I posed [this question](https://github.com/exercism/ruby/pull/1289#discussion_r790748112), which then led `kotp` to lead a discussion on whether it was right to make the change at all, and they decided to drop the PR in the end. If I'd messaged directly saying "I don't think this is good" etc, I'd have felt like I was undermining `kotp`'s review, and potentially confused the author as to who's advice to follow. I have to be like triple-careful of this in my role in Exercism, but I think it's good practice in general.

My final thought is that personally I always try to ensure I don't add work to maintainers load, so I tend to ask myself whether my review is duplicating work that will happen anyway, or adding something useful. For most of Exercism I generally don't make comments on PRs because I think that when the maintainer does the full review, they'll think of (and make) those comments anyway, so I'm not really saving them work, and actually giving them more to read. Where I do tend to comment is on areas where I think something's been missed, there's some wrong first-principle logic, or where I have some insight that others might not have. 

So @IsaacG, I think your input on changes to exercises is incredibly valuable to the Python track. I 100% want you to feel empowered to say "I think this is a bad change based on X reasons due to my interactions with students" - we need to hear that. And sometimes you might see a new exercise, and see a glaring problem too. Definitely speak up in those places. But probably as a general rule of thumb, posting an overarching comment on the PR has loads more value than doing a formal review of the PR, which can add noise/confusion (largely due to the way GH's UI works). 

**So as a TL;DR: I agree with Bethany's comment of _"That doesn't mean no comments or discussion on them -- but that does mean that formally reviewing [something] is discouraged"_ but for all the complex reasons listed above. And I'd say that I'd invert the tone too, adding "Please post comments on PRs where you think there are mistakes being made or fundamental improvements that could be made, once a maintainer has had chance to review".** I hope that wall of text adds some value!

File: concepts/classes/about.md
Comment (BethanyG) [inline concepts/classes/about.md:292]:
```suggestion
In previous concept exercises and practice exercise stubs, you will have seen the `pass` keyword used within the body of functions in place of actual code.

```

File: exercises/concept/ellens-alien-game/.docs/hints.md
Comment (BethanyG) [inline exercises/concept/ellens-alien-game/.docs/hints.md:5]:
```suggestion
- Remember that object methods are always passed `self` as the first parameter.
```

File: exercises/concept/ellens-alien-game/.docs/hints.md
Comment (BethanyG) [inline exercises/concept/ellens-alien-game/.docs/hints.md:7]:
```suggestion
- Instance variables are unique to the class instance (object) that possesses them.
```

File: exercises/concept/ellens-alien-game/.docs/hints.md
Comment (BethanyG) [inline exercises/concept/ellens-alien-game/.docs/hints.md:12]:
```suggestion
- Remember that object methods are always passed `self` as the first parameter.
```

File: exercises/concept/ellens-alien-game/.docs/hints.md
Comment (BethanyG) [inline exercises/concept/ellens-alien-game/.docs/hints.md:16]:
```suggestion
- Remember that object methods are always passed `self` as the first parameter.
```

File: exercises/concept/ellens-alien-game/.docs/hints.md
Comment (BethanyG) [inline exercises/concept/ellens-alien-game/.docs/hints.md:21]:
```suggestion
- Remember that object methods are always passed `self` as the first parameter.
```

File: exercises/concept/ellens-alien-game/.docs/hints.md
Comment (BethanyG) [inline exercises/concept/ellens-alien-game/.docs/hints.md:25]:
```suggestion
- Remember that object methods are always passed `self` as the first parameter.
```

File: exercises/concept/ellens-alien-game/.docs/hints.md
Comment (BethanyG) [inline exercises/concept/ellens-alien-game/.docs/hints.md:26]:
```suggestion
- If you wish to 'pass' on completing a method, Python has a keyword for that.
```

File: exercises/concept/ellens-alien-game/.docs/hints.md
Comment (BethanyG) [inline exercises/concept/ellens-alien-game/.docs/hints.md:36]:
```suggestion
- Unpacking what is _inside_ the tuple might be closer to what you need.
```

File: exercises/concept/ellens-alien-game/.docs/introduction.md
Comment (BethanyG) [inline exercises/concept/ellens-alien-game/.docs/introduction.md:125]:
```suggestion
        # Note that we are unpacking the tuple argument into two separate instance variables.
```

File: exercises/concept/ellens-alien-game/.docs/introduction.md
Comment (BethanyG) [inline exercises/concept/ellens-alien-game/.docs/introduction.md:153]:
```suggestion
class My_Class:
    number = 5
    string = "Hello!"
    
    #Class constructor.
    def __init__(self, location):
        # Instance properties
        self.location_x = location[0]
        self.location_y = location[1]
        
    #Instance method. Note "self" as first parameter.
    def change_location(self, amount):
        self.location_x += amount
        self.location_y += amount
        return self.location_x, self.location_y
```

File: exercises/concept/ellens-alien-game/.meta/config.json
Comment (BethanyG) [inline exercises/concept/ellens-alien-game/.meta/config.json:5]:
```suggestion
  "contributors": ["DjangoFett", "kotp"],
```

File: concepts/classes/.meta/config.json
Comment (BethanyG) [inline concepts/classes/.meta/config.json:3]:
```suggestion
  "authors": ["bethanyg", "DjangoFett", "PaulT89"],
```

**Review by BethanyG — APPROVED** [review]
@PaulT89  -- I think at loooooong last, _**we are there!!*_ 🌟 🌈 
I've approved this -- but will put a hold on it and wait for you to give it one last read-through and check before we merge it in.  😄    **Thank you so much** for your excellent work on this!

**Comment by PaulT89** [comment]
Okay, I'm happy.  Carefully reading through it, I found a bunch of spelling mistakes - my spell-checker didn't pick them up because they were all in the comments within the code-blocks - and I think I got them all (but, admittedly, I might have missed some, as most were in both `about.md` and the exercise `introduction.md`).  Added 3 links to `links.json`, re-read the `instructions.md` & `hints.md` and fixed a few clumsy sentences and inconsistencies.  That's about it really.  Good to go. 👍

**Comment by PaulT89** [comment]
Oh, and what happens to the "wip" tag in the main `config.json` file?  Does that need to come off or do we keep that on for now?

**Comment by BethanyG** [comment]
> Oh, and what happens to the "wip" tag in the main config.json file? Does that need to come off or do we keep that on for now?

Per our earlier discussion on this PR, we're going to keep the status `WIP` for now as we merge.  😄 

@IsaacG -- Now would be the time (_if you have time and are willing_) to go through the exercise, create a solution, and put together an issue on what you think needs changing or adding.  The original issue for this exercise is [here](https://github.com/exercism/python/issues/2349), which includes the topics covered, etc.  I think it might be cleanest to do a brand-new "feedback on Ellens Alien Game" issue linking this PR and the original.

**Comment by kotp** [comment]
I think we can remove the "Do Not Merge" Label, or someone might think to revert this as a mistaken merge.

**Pull request merged**
